pkg/bpfassets/attacher/libbpf_attacher.go

```diff
@@ -0,0 +1,304 @@
+	arch := runtime.GOARCH
+	libbpfObjectFilePath, err = getLibbpfObjectFilePath(arch)
+	if err == nil {
+		libbpfModule, err = bpf.NewModuleFromFile(libbpfObjectFilePath)
+	}
+	if err != nil {
+		return nil, fmt.Errorf("failed to load module: %v", err)
+	}
+
+	err = libbpfModule.BPFLoadObject()
+
+	// attach sched_switch tracepoint to kepler_trace function
+	prog, err := libbpfModule.GetProgram("kepler_trace")
+	if err != nil {
+		return libbpfModule, fmt.Errorf("failed to get kepler_trace: %v", err)
+	} else {
+		_, err = prog.AttachTracepoint("sched:sched_switch")
+		if err != nil {
+			return libbpfModule, fmt.Errorf("failed to attach sched/sched_switch: %v", err)
+		}
+	}
+
+	// attach softirq_entry tracepoint to kepler_irq_trace function
+	irq_prog, err := libbpfModule.GetProgram("kepler_irq_trace")
+	if err != nil {
+		klog.Infof("failed to get kepler_irq_trace: %v", err)
+		// disable IRQ metric
+		config.ExposeIRQCounterMetrics = false
+	} else {
+		_, err = irq_prog.AttachTracepoint("irq:softirq_entry")
+		if err != nil {
+			klog.Infof("failed to attach irq/softirq_entry: %v", err)
+			// disable IRQ metric
+			config.ExposeIRQCounterMetrics = false
+		}
+	}
+
+	// set counters
+	Counters = libbpfCounters
+
+	// attach performance counter fd to BPF_PERF_EVENT_ARRAY
+	for arrayName, counter := range Counters {
+		bpfPerfArrayName := arrayName + BpfPerfArrayPrefix
+		bpfMap, perfErr := libbpfModule.GetMap(bpfPerfArrayName)
+		if perfErr != nil {
+			klog.Infof("failed to get perf event %s: %v\n", bpfPerfArrayName, perfErr)
+			continue
+		}
+		perfErr = unixOpenPerfEvent(bpfMap, counter.EvType, counter.EvConfig)
+		if perfErr != nil {
+			// some hypervisors don't expose perf counters
+			klog.Infof("failed to attach perf event %s: %v\n", bpfPerfArrayName, perfErr)
+			counter.enabled = false
+
+			// if any counter is not enabled, we need disable HardwareCountersEnabled
+			HardwareCountersEnabled = false
+		}
+	}
+
+	klog.Infof("Successfully load eBPF module from libbpf object")
+	return libbpfModule, nil
+}
+
+func detachLibbpfModule() {
+	unixClosePerfEvent()
+	if libbpfModule != nil {
+		libbpfModule.Close()
+		libbpfModule = nil
+	}
+}
+
+func libbpfCollectProcess() (processesData []ProcessBPFMetrics, err error) {
+	processesData = []ProcessBPFMetrics{}
+	if libbpfModule == nil {
+		// nil error should be threw at attachment point, return empty data
+		return
+	}
+	var processes *bpf.BPFMap
+	processes, err = libbpfModule.GetMap(TableProcessName)
+	if err != nil {
+		return
+	}
+	processKeySize := int(unsafe.Sizeof(uint64Key))
+	iterator := processes.Iterator(processKeySize)
+	var ct ProcessBPFMetrics
+	valueSize := int(unsafe.Sizeof(ProcessBPFMetrics{}))
+	keys := []uint32{}
+	retry := 0
+	next := iterator.Next()
+	for next {
+		keyBytes := iterator.Key()
+		key := ByteOrder.Uint32(keyBytes)
+		data, getErr := processes.GetValue(key, valueSize)
+		if getErr != nil {
+			retry += 1
+			if retry > maxRetry {
```

Comment:
@AdrianHammond hit an issue with maxRetry and gets many error message

Reply:
Put my response here: https://github.com/sustainable-computing-io/kepler/pull/791#issuecomment-1635121601